core/treepath/treepath.go

```diff
@@ -0,0 +1,173 @@
+// Package treepath provides functionality for manipulating treepaths.
+//
+// Treepaths are string-encodings of paths through trees.
+//
+// Treepaths come originially from
+// https://github.com/Kashomon/glift-core/blob/master/src/rules/treepath.js
+package treepath
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/otrego/clamshell/core/game"
+)
+
+// A Treepath is a list of variations that says how to travel through a tree of
+// moves. And has two forms a list version and a string version. First, the
+// list-version:
+//
+//    [0,1,0]
+//
+// Means we will first take the 0th variation, then we will take the 1ist
+// variation, and lastly we will take the 0th variation again. For
+// convenience, the treepath can also be specified by a string, which is where
+// the fun begins. At its simpliest,
+//
+//    [0,0,0] becomes 0.0.0
+//
+// but there are a couple different string short-hands that make using treepaths
+// a little easier
+//
+//    0.1:2   Take the 0th variation, then repeat taking the 1st varation twice
+//
+// There are two types of treepaths discussed below -- A *treepath fragment*
+// (which is what we have been describing) and an *initial treepath*.
+//
+// ## Treepath
+//
+// Treepaths say how to get from position n to position m.  Thus the numbers are
+// always variations except in the case of AxB syntax, where B is a multiplier
+// for a variation.
+//
+// This is how fragment strings are parsed:
+//
+//    0             becomes [0]
+//    1             becomes [1]
+//    53            becomes [53] (the 53rd variation)
+//    2.3           becomes [2,3]
+//    0.0.0.0       becomes [0,0,0]
+//    0:4           becomes [0,0,0,0]
+//    1:4           becomes [1,1,1,1]
+//    1.2:1.0.2:3   becomes [1,2,0,2,2,2]
+type Treepath []int
+
+type parseState int
+
+const (
+	// variationState means we are looking for numbers for the variation number. Can
+	// either transition to either separator (flush) or repeat (flush and then
+	// repeat N times
+	variationState parseState = iota
+	// repeat means we repeat the previous variation 'n' times.
+	repeatState
+)
+
+// Parse parses a treepath from a string to an array of ints.
+//
+// Example:
+// * Parse("2.3")         => [2,3]
+// * Parse("1.2:1.0.2:3") => [1,2,0,2,2,2]
+func Parse(path string) (Treepath, error) {
+	out := Treepath{}
+	curState := variationState
+	buf := &strings.Builder{}
+	prevChar := 0
+
+	convertBuffer := func(idx int) (int, error) {
+		if buf.Len() == 0 {
+			return 0, fmt.Errorf("error parsing fragment %q at index %d: separator '.' must be proceeded by digit", path, idx)
```

Comment:
For the error message here, it could be not a `.` but a `:` or EOF, right? Think it's worth the complexity to have correct error messages?

Reply:
Done